hubblestack_nova/modules/openssl.py

```diff
@@ -0,0 +1,283 @@
+                success.append({tag: description})
+                tags_descriptions.add((tag, description))
+
+        control_reasons = set()
+
+        for tag_data in ret['Controlled']:
+            tag = tag_data['tag']
+            description = tag_data.get('description')
+            control_reason = tag_data.get('control', '')
+            if (tag, description, control_reason) not in control_reasons:
+                tag_dict = {'description': description,
+                            'control': control_reason}
+                controlled.append({tag: tag_dict})
+                control_reasons.add((tag, description, control_reason))
+
+        ret['Controlled'] = controlled
+        ret['Success'] = success
+        ret['Failure'] = failure
+
+    if not ret['Controlled']:
+        ret.pop('Controlled')
+
+    return ret
+
+
+def _merge_yaml(ret, data):
+    if 'openssl' not in ret:
+        ret['openssl'] = []
+    for key, val in data.get('openssl', {}).iteritems():
+        ret['openssl'].append({key: val})
+    return ret
+
+
+def _get_tags(data):
+    ret = {}
+    for audit_dict in data.get('openssl', {}):
+        for audit_id, audit_data in audit_dict.iteritems():
+            tags_dict = audit_data.get('data', {})
+            tag = tags_dict.pop('tag')
+            if tag not in ret:
+                ret[tag] = []
+            formatted_data = copy.deepcopy(tags_dict)
+            formatted_data['tag'] = tag
+            formatted_data['module'] = 'openssl'
+            formatted_data.update(audit_data)
+            formatted_data.pop('data')
+            ret[tag].append(formatted_data)
+    return ret
+
+
+def _check_x509(x509=None, notBefore=0, notAfter=0, fail_if_notBefore=False):
+    if not x509:
+        return (False, 'No certificate to be checked')
+    if x509.has_expired():
+        return (False, 'The certificate is expired')
+
+    stats = _get_x509_days_left(x509)
+
+    if notAfter >= stats['notAfter']:
+        return (False, 'The certificate will expire in less then {0} days'.format(notAfter))
+    if notBefore <= stats['notBefore']:
+        if notBefore == 0 and fail_if_notBefore:
+            return (False, 'The certificate is not yet valid ({0} days left until it will be valid)'.format(stats['notBefore']))
+        return (False, 'The certificate will be valid in more then {0} days'.format(notBefore))
+
+    return (True, '')
+
+
+def _load_x509(source, port=443, from_file=False):
+    if not from_file:
+        x509 = _load_x509_from_endpoint(source, port)
+    else:
+        x509 = _load_x509_from_file(source)
+    return x509
+
+
+def _load_x509_from_endpoint(server, port=443):
+    try:
+        cert = ssl.get_server_certificate((server, port))
+    except Exception:
+        cert = None
+    if not cert:
+        return None
+
+    try:
+        x509 = OpenSSL.crypto.load_certificate(OpenSSL.crypto.FILETYPE_PEM, cert)
+    except OpenSSL.crypto.Error:
```

Comment:
I wonder if we should log these errors at debug log level or something, so people can more easily track down what's going wrong?